cmd/scollector/collectors/awsBilling.go

```diff
@@ -0,0 +1,445 @@
+	productCodes := []string{}
+	for _, line := range b.LineItems {
+		tags := make(opentsdb.TagSet)
+		//Keep track of the unique product codes we've seen so we can
+		//send metadata for them later
+		if !stringInSlice(line.ProductCode, productCodes) {
+			productCodes = append(productCodes, line.ProductCode)
+		}
+		//Add the host tag, and delete the host tag if it's blank
+		if host, present := tags["host"]; !present {
+			tags["host"] = util.Hostname
+		} else if host == "" {
+			delete(tags, "host")
+		}
+		//Suppressing AccountID for now as we only have one account
+		//tags["accountid"] = line.UsageAccountID
+		tags["operation"] = line.Operation
+		metricCode := strings.ToLower(line.ProductCode)
+		metricCode = strings.Replace(metricCode, "amazon", "", 1)
+		//Usage Type is meaningless for R53, they're all 'DNS-Queries'
+		if line.ProductCode != "AmazonRoute53" {
+			tags["usagetype"] = line.UsageType
+		}
+		//Use the zone name if present for Route 53 data
+		if line.ProductCode == "AmazonRoute53" && line.Route53Zone != nil && *line.Route53Zone.Name != "" {
+			tags["resourceid"] = strings.ToLower(*line.Route53Zone.Name)
+		} else if line.ResourceID != "" {
+			//Remove pesky invalid characters that show up. Probably should redo this as a regex
+			tags["resourceid"] = strings.ToLower(strings.Replace(strings.Replace(strings.Replace(strings.Replace(line.ResourceID, "/", "-", -1), ":", "-", -1), "[", "", -1), "]", "", -1))
+		}
+		//For our denormalized metrics, just keep the host tag, and the required tag
+		//Some entries do not have a resource ID (e.g. access denied errors, or bucket does not exist, for obvious reasons)
+		tagsByResource := opentsdb.TagSet{"host": tags["host"]}
+		if resid, present := tags["resourceid"]; present {
+			tagsByResource["resourceid"] = resid
+		}
+		tagsByOperation := opentsdb.TagSet{"host": tags["host"], "operation": tags["operation"]}
+		descCosts := fmt.Sprintf("Usage costs for Amazon %v. Datapoints represent costs for a full hour. Data typically lags by 24 hours.", metricCode)
+		descUsage := fmt.Sprintf("Usage volume for Amazon %v. Datapoints represent a full hour of usage. Data typically lags by 24 hours.", metricCode)
+		descUsageRes := fmt.Sprintf("Usage volume for Amazon %v, denormalized with only a resource tag. Datapoints represent a full hour of usage. Data typically lags by 24 hours.", metricCode)
+		descUsageOps := fmt.Sprintf("Usage volume for Amazon %v, denormalized with only an operation tag. Datapoints represent a full hour of usage. Data typically lags by 24 hours.", metricCode)
+		//Glacier bills by day, not by hour, for storage, so we need to fill in some fake datapoints
+		if line.ProductCode == "AmazonGlacier" && line.Operation == "Storage" {
+			for i := 0; i < 24; i++ {
+				AddTS(md, fmt.Sprintf("aws.billing.%v.cost", metricCode), line.UsageEndDate.Add(time.Duration(-1*i)*time.Hour).Unix(), line.UnblendedCost/24, tags, metadata.Gauge, metadata.Count, descCosts)
```

Comment:
When you have something like line.UsageEndDate.Add(time.Duration(-1*i)*time.Hour).Unix(), don't call it multiple times, save the result in a var and then pass that.